Several fixes to make onboarding and some other features work properly.

begedin opened the pull request:
- UserRoleController was missing index and show endpoints. I added those, but we still need to add tests
- Added routes to support above actions
- Added has_many user-skills to skill view, but I'm not absolutely sure we need that. The ember behavior on the skill onboarding page is weird. We should look into making it simpler.
- Fixed fitlering of emails for unauthenticated user. This was an issue introduced by me today. I was under the impression that conn will get a "current_user" assign when authenticated, but the current user is actually assigned to the "data" key. I changed the method signature in UserView, but we need to update associated tests and add a test on the user controller, best under the show group, to ensure an authenticated user will get an email. Apparently, the view test wasn't enough.
- **Very important** - Made it so our pipeline always loads the current user if a token is present. The endpoints that do explicitly require authentication should still support requests from authenticated users.
- [ ] Add tests for UserRoleController #index and #show. Use UserSkillController as example
- [ ] Look into making ember behavior on the skill onboarding page simpler
- [x] Update broken tests on the user view
- [x] Add test for UserController.show, where, when authenticated, the response has the email field properly populated.

JoshSmith commented:
Do you know if there's any real way to test this? I'm not sure if that's realistic or not, but thought I'd ask.

ptrikutam commented:
Good question. I don't really know, actually. [A quick google search](https://stackoverflow.com/questions/1339416/unit-testing-bash-scripts) reveals that there are things out there that do that. Perhaps something that we could do in a separate PR?

ptrikutam commented:
Looks like the solutions aren't great for unit testing shell scripts. Might add more complexity than is worth it to test a relatively simple script so far.

begedin committed:
Several fixes to make onboarding and some other features work properly.

- UserRoleController was missing index and show endpoints. I added those, but we still need to add tests
- Added routes to support above actions
- Fixed fitlering of emails for unauthenticated user. This was an issue introduced by me today. I was under the impression that conn will get a "current_user" assign when authenticated, but the current user is actually assigned to the "data" key. I changed the method signature in UserView, but we need to update associated tests and add a test on the user controller, best under the show group, to ensure an authenticated user will get an email. Apparently, the view test wasn't enough.
- Fix tests, make it so routes that do not require authentication, still load the authenticated user if information exists

begedin commented:
What's next?

I think we should simplify the user view and not include the "through" relationships (skills, roles, categories). Instead, we should only include "user-skills", "user-roles", "user-categories". The client is more then capable to fetch the other stuff. Ember might need to be updated as well.

joshsmith commented on web/views/skill_view.ex:
```diff
@@ -5,5 +5,5 @@ defmodule CodeCorps.SkillView do
 
   has_many :role_skills, serializer: CodeCorps.RoleSkillView
-  has_many :roles, serializer: CodeCorps.RoleView
+  has_many :user_skills, serializer: CodeCorps.UserSkillView
```
Do we really want all the `user_skills` across the app loaded for each `Skill`? I wouldn't think so.

I can see adding it for a `User`, but not for a `Skill`.

joshsmith reviewed (changes requested):
Not sure how I feel about the changes made here.

joshsmith commented:
@begedin I see you commented that you weren't sure about including user skill in skill. I removed that. It's not necessary after all and appears to still work on Ember side.
